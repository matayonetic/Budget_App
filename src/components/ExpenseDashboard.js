import React from "react";

// Dashboard
const ExpenseDashboard = () => {
  return (
    <div>
      <h2>Home Page</h2>
    </div>
  );
};

export default ExpenseDashboard