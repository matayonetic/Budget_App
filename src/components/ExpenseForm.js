import React from "react";
import moment from "moment";
import "react-dates/initialize";
import { SingleDatePicker } from "react-dates";

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.changeDescription = this.changeDescription.bind(this);
    this.changeAmount = this.changeAmount.bind(this);
    this.state = {
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
      amount: props.expense ? (props.expense.amount / 100).toString() : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: "",
    };
  }

  // Change Description
  changeDescription(e) {
    const description = e.target.value;
    this.setState(() => {
      return {
        description,
      };
    });
  }

  // Change Amount
  changeAmount(e) {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  }

  // Change Note
  changeNote = (e) => {
    const note = e.target.value;
    this.setState(() => ({
      note,
    }));
  };  

  // Change Date
  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };

  // Focus Change
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };

  // Submit Form
  onSubmit = (e) => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => {
        return {
          error: "Please check description or amount",
        };
      });
    } else {
      this.setState(() => ({
        error: "",
      }));
      this.props.formData({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note,
      });
    }
  };

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Enter description"
            autoFocus
            value={this.state.description}
            onChange={this.changeDescription}
          />
          <br />
          <br />
          <input
            type="number"
            placeholder="Enter amount"
            value={this.state.amount}
            onChange={this.changeAmount}
          />
          <br />
          <br />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <br />
          <br />
          <textarea
            placeholder="Add a note(optional)..."
            value={this.state.note}
            onChange={this.changeNote}
          ></textarea>
          <br />
          <br />
          <button>Add Expense</button>
        </form>
      </div>
    );
  }
}
