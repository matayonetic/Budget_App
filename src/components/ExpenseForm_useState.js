import React, { useState } from "react";
import moment from "moment";
import "react-dates/initialize";
import { SingleDatePicker } from "react-dates";
import "react-dates/lib/css/_datepicker.css";

const ExpenseForm = (props) => {
  //
  // Tracking Form Inputs
  const [calenderFocused, setFocus] = useState(false);
  const [error, setError] = useState("");
  const [note, setNote] = useState(props.expense ? props.expense.note : "");

  const [amount, setAmount] = useState(
    props.expense ? (props.expense.amount / 100).toString() : ""
  );

  const [createdAt, setDate] = useState(moment());

  const [description, setDescription] = useState(
    props.expense ? props.expense.description : ""
  );

  // Change Description
  const changeDescription = (e) => {
    setDescription(e.target.value);
  };

  // Change Note
  const changeNote = (e) => {
    setNote(e.target.value);
  };

  // Change Amount
  const changeAmount = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      setAmount(amount);
    }
  };

  // Change Date
  const onDateChange = (createdAt) => {
    if (createdAt) {
      setDate(createdAt);
    }
  };

  // Change Focus
  const onFocusChange = ({ focused }) => {
    setFocus(focused);
  };

  // Submit Form
  const onSubmit = (e) => {
    e.preventDefault();
    if (!description || !amount) {
      setError("Please check description or amount");
    } else {
      setError("");

      // Destructure "expense" argument
      props.formData({
        description,
        amount: parseFloat(amount, 10) * 100,
        createdAt: createdAt.valueOf(),
        note,
      });
    }
  };

  return (
    <div>
      {error && <p>{error}</p>}
      <form onSubmit={onSubmit}>
        <input
          placeholder="Description"
          type="text"
          value={description}
          onChange={changeDescription}
          autoFocus
        />
        <br />
        <br />
        <input
          placeholder="Amount"
          type="number"
          value={amount}
          onChange={changeAmount}
        />
        <br />
        <br />
        <SingleDatePicker
          date={createdAt}
          onDateChange={onDateChange}
          focused={calenderFocused}
          onFocusChange={onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <br />
        <br />
        <textarea
          placeholder="Additional notes"
          value={note}
          onChange={changeNote}
        ></textarea>
        <br />
        <br />
        <button>Add Expense</button>
      </form>
    </div>
  );
};

export default ExpenseForm;
