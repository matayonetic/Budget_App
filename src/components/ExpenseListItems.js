import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import moment from "moment"
import numeral from "numeral"

export const ExpenseListItems = ({ id, description, amount, note, createdAt }) => {
  return (
    <div>
      <Link to={`/edit/${id}`}>        
        <h3>{description}</h3>
      </Link>
      
      <p>Amount: {numeral(amount / 100).format("$0,0.00")}</p>
      <p>Created At: {moment(createdAt).format('MMMM Do, YYYY' )}</p>
      <p> Note: {note}</p>
    </div>
  );
};

export default connect()(ExpenseListItems);
