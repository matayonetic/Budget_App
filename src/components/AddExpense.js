import React from "react";
import { useNavigate } from "react-router-dom";
import { connect } from "react-redux";

import { addExpense } from "../actions/expensesGen";
import ExpenseForm from "./ExpenseForm.js";

// Class
class AddExpenseClass extends React.Component {
  //
  addExpense = (expense) => {
    const { navigate } = this.props;
    this.props.addExpenseData(expense);
    navigate("/");
  };

  render() {
    return (
      <div>
        <h2>Add Expense</h2>
        <ExpenseForm formData={this.addExpense} />
      </div>
    );
  }
}

// Add Expense
export const AddExpense = (props) => {
  let navigate = useNavigate();
  return <AddExpenseClass {...props} navigate={navigate} />;
};

// Map Dispatch to Props
const mapDispatchToProps = (dispatch) => ({
  addExpenseData: (expense) => {
    dispatch(addExpense(expense));
  },
});

// Default Export
export default connect(undefined, mapDispatchToProps)(AddExpense);
