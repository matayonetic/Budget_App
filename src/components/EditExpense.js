import React from "react";
import ExpenseForm from "./ExpenseForm";
import { startEditExpense, startRemoveExpense } from "../actions/expensesGen";
import { useNavigate, useParams } from "react-router-dom";
import { useSelector, useDispatch } from "../react-redux-hooks";

// Edit Expense Component
const EditExpense = () => {
  //
  const { id } = useParams();
  const navigate = useNavigate();
  const dispatch = useDispatch();

  // Expense
  const expense = useSelector((state) =>
    state.expenses.find((expense) => expense.id === id)
  );

  // Edit Expense
  const editMyExpense = (expense) => {
    dispatch(startEditExpense(id, expense));
    navigate("/dashboard");
  };

  // Remove Expense
  const removeMyExpense = () => {
    dispatch(startRemoveExpense({ id }));
    navigate("/dashboard");
  };

  // Render
  return (
    <div>
      <div className="page-header">
        <div className="content-container">
          <h2 className="page-header__title">Edit Expense</h2>
        </div>
      </div>
      <div className="content-container">
        <ExpenseForm expense={expense} formData={editMyExpense} />
        <button className="button button__secondary" onClick={removeMyExpense}>
          Remove Expense
        </button>
      </div>
    </div>
  );
};

export default EditExpense;
