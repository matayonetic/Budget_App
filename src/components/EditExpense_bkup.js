import React from "react";
import ExpenseForm from "./ExpenseForm";
import { editExpense, removeExpense } from "../actions/expensesGen";
import { useNavigate, useParams } from "react-router-dom";
import { useSelector, useDispatch } from "react-redux";

class EditExpenseClass extends React.Component {
  //
  // Edit Expense
  editExpense = (expense) => {
    const { navigate } = this.props;
    const id = this.props.id;
    this.props.editExpenseData(id, expense);
    navigate("/");
  };

  // Remove Expense
  removeExpense = (id) => {
    const { navigate } = this.props;
    id = this.props.id;
    this.props.removeExpenseData(id);
    navigate("/");
  };

  // Render
  render() {
    return (
      <div>
        <h2>Edit Expense</h2>
        <ExpenseForm expense={this.props.expense} formData={this.editExpense} />
        <p>
          <button onClick={this.removeExpense}>Remove</button>
        </p>
      </div>
    );
  }
}

// Edit Expense Component
const EditExpense = (props) => {
  //
  // Expense ID
  const { id } = useParams();

  // Expense
  const expense = useSelector((state) =>
    state.expenses.find((expense) => expense.id === id)
  );

  // Navigation
  let navigate = useNavigate();

  // Dispatch
  const dispatch = useDispatch();

  // Dispatch Edit
  const editExpenseData = (id, expense) => {
    dispatch(editExpense(id, expense));
  };

  // Dispatch Remove
  const removeExpenseData = (id) => {
    dispatch(removeExpense({ id }));
  };

  // Render
  return (
    <EditExpenseClass
      {...props}
      id={id}
      expense={expense}
      navigate={navigate}
      editExpenseData={editExpenseData}
      removeExpenseData={removeExpenseData}
    />
  );
};

// const matchDispatchToProps = (dispatch) => {
//   return {
//     editExpenseData: (id, expense) => {
//       dispatch(editExpense(id, expense));
//     },
//     removeExpenseData: (id) => {
//       dispatch(removeExpense({ id }));
//     },
//   };
// };

export default EditExpense;
