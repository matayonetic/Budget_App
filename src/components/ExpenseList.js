import React from "react";
import { useSelector } from "../react-redux-hooks";
import ExpenseListItems from "./ExpenseListItems";
import getVisibleExpenses from "../selectors/getVisibleExpenses";

const ExpenseList = () => {
  const expenses = useSelector((state) =>
    getVisibleExpenses(state.expenses, state.filters)
  );
  return (
    <div>
      <h2>Expense List</h2>
      {expenses.length > 0 ? (
        expenses.map((expense) => {
          return <ExpenseListItems key={expense.id} {...expense} />;
        })
      ) : (
        <p>No data found</p>
      )}
    </div>
  );
};

export default ExpenseList;
