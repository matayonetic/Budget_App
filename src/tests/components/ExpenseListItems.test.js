import React from "react";
import { shallow } from "enzyme";
import { ExpenseListItems } from "../../components/ExpenseListItems";
import expenses from "../fixtures/expenses";

// Render ExpenseListItems
test("Should render expense list items", () => {
  const expense = expenses[0];  
  const wrapper = shallow(<ExpenseListItems {...expense} />);
  expect(wrapper).toMatchSnapshot();
});
