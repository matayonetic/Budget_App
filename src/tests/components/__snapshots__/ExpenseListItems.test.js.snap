// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render expense list items 1`] = `
<div>
  <Link
    to="/edit/abc1"
  >
    <h3>
      Food
    </h3>
  </Link>
  <p>
    Amount: 
    $0.01
  </p>
  <p>
    Created At: 
    December 31st, 1969
  </p>
  <p>
     Note: 
    a
  </p>
</div>
`;
